wad representation

ozgunozerk opened the pull request:
<!-- Thank you for your interest in contributing to OpenZeppelin! -->

<!-- Consider opening an issue for discussion prior to submitting a PR. -->
<!-- New features will be merged faster if they were first discussed and designed with the team. -->

Partially Fixes #438 

<!-- Describe the changes introduced in this pull request. -->
<!-- Include any context necessary for understanding the PR's purpose. -->

#### PR Checklist

<!-- For the PRs that introduce new features, all of the following must be complete. -->
<!-- Feel free to submit a PR or Draft PR even if some items are pending. -->
<!-- Some of the items below may not apply to your case (for example: fixing a typo). -->

- [x] Tests
- [x] Documentation

brozorec reviewed (commented):
1. Document truncation behavior consistently
Similarly to `to_integer()`:
    - Add to to_token_amount(): "Truncates toward zero when scaling down (token_decimals < 18)."
    - Add to from_token_amount(): "Truncates toward zero when scaling down (token_decimals > 18)."
    - Add to checked_mul(): "Result is truncated toward zero after division by WAD_SCALE."
    - Add to checked_div(): "Result is truncated toward zero."
2. Negatives are supported so `abs()` can be useful.
3. Can be helpful to provide:
```
impl Wad {
    pub const ZERO: Wad = Wad(0);
    pub const ONE: Wad = Wad(WAD_SCALE);
}
```
4. Precision loss: is this inevitable?
```
    #[test]
    fn test_precision_loss_operation_order() {
        let e = Env::default();

        let principal = Wad::from_integer(&e, 1000);
        let rate = Wad::from_raw(55_000_000_000_000_000);
        let time_fraction = Wad::from_raw(8_333_333_333_333_333);

        let interest_method1 = principal * rate * time_fraction;

        let rate_time = rate * time_fraction;
        let interest_method2 = principal * rate_time;

        let diff = interest_method1.raw() - interest_method2.raw();

        assert!(diff > 0);
    }

    #[test]
    fn test_precision_loss_compound_interest() {
        let e = Env::default();

        let principal = Wad::from_integer(&e, 100);
        let growth_factor = Wad::from_raw(1_010_000_000_000_000_000);

        let mut amount = principal;
        for _ in 0..10 {
            amount = amount * growth_factor;
        }

        let expected = Wad::from_raw(110_462_212_541_000_000_000);

        let diff = amount.raw() - expected.raw();

        assert!(diff > 120_451_000);
    }

    #[test]
    fn test_precision_loss_cross_multiplication() {
        let a = Wad::from_raw(10_000_000_000_000_000_007);
        let b = Wad::from_raw(3_000_000_000_000_000_000);

        let quotient = a / b;
        assert_eq!(quotient.raw(), 3_333_333_333_333_333_335);

        let reconstructed = quotient * b;
        assert_eq!(reconstructed.raw(), 10_000_000_000_000_000_005);

        let lost = a.raw() - reconstructed.raw();
        assert_eq!(lost, 2);
    }
```
5. Add additional checks for large decimals
```
    #[test]
    #[should_panic(expected = "attempt to multiply with overflow")]
    fn test_token_decimals_not_validated() {
        let e = Env::default();

        let amount = 1_000_000;
        let invalid_decimals = 57u8;

        let _ = Wad::from_token_amount(&e, amount, invalid_decimals);
    }
```

ozgunozerk commented:
> Add additional checks for large decimals

that's good, added the test

> Negatives are supported so abs() can be useful.

I had `Neg` and `abs` implementation, but removed them. Can add once more, sure.

> Can be helpful to provide:
> impl Wad {
>    pub const ZERO: Wad = Wad(0);
>    pub const ONE: Wad = Wad(WAD_SCALE);
>}

Hmm, maybe `zero`, but definitely not `ONE`, because it will create ambiguity. Is it `1` token (full token), or `1` smallest unit for the token? I don't think this is a good idea... And I also had `zero` in the beginning, but since `one` does not make much sense, I also removed the `zero` for consistency. They can easily create `zero` from `from_integer(0)`. I don't see a huge benefit of having this tbh.

> Document truncation behavior consistently

good advice, will do!

> Precision loss: is this inevitable?

I think yes, but will spend more mental effort on it just in case.

ozgunozerk commented on packages/contract-utils/src/math/test.rs:
```diff
@@ -672,3 +673,427 @@ mod test_i256_errors {
+        let invalid_decimals = 57u8;
+
+        let _ = Wad::from_token_amount(&e, amount, invalid_decimals);
+    }
```
Good idea, will do that!

brozorec reviewed (approved):
I'm still not feeling 100% convinced about the operator overloading, but acknowledging  the arguments about this design choice. One last thing, there's one miss from the test cov that can be easily covered.

Great job 🙌
